Ambient User Guide - L7 Waypoint Proxy Layer Sub Guide (#13954)
istio/istio.io: content/en/docs/ops/ambient/usage/waypoint/index.md

---
title: L7 - waypoint proxy Layer
description: User guide for L7 Processing.
weight: 2
owner: istio/wg-networking-maintainers
test: n/a
---

{{<tip>}}

> Before you start with this guide, please make sure that you have already read the Ztunnel Networking sub-guide and have a basic understanding of Istio Ambient Mesh. This guide assumes that you have already completed the Installation, set up the ztunnel and enabled mTLS.

Reviewer: possible to have link in tip?

would like to have link to Ztunnel Networking sub-guide or Installation, set up the ztunnel and enabled mTLS

> If you have not yet done these things, please go back to the Ztunnel Networking sub-guide and follow the instructions before proceeding with this guide.

Reviewer 2: Could remove this line .. you are already stating above and below that ztunnel is a pre-requisite

Author: sure, done!

Reviewer: pre-requisites would be nice, especially if links won't work in tip section.

Once you have completed the Ztunnel Networking sub-guide, you will be ready to start setting up the L7 Waypoint Proxy layer in Istio Ambient Mesh.

{{</tip>}}

## Introduction 

> This guide provides instructions on how to set up and use the L7 Waypoint Proxy layer in Istio Ambient Mesh. It assumes that you have already read the Ztunnel Networking sub-guide and have a basic understanding of Istio Ambient Mesh.

Reviewer: +1 ^^

> L7 traffic routing is based on L4 with the addition of the Waypoint proxy, which is more complex to handle in Envoy. We can also create HPAs to scale it dynamically.

Reviewer 3: Might consider rewording to the following:  L7 traffic routing is handled via the Waypoint proxy.  The waypoint proxy is currently based on Envoy.  The waypoint proxy can be deployed to scale dynamically using HPAs.

Author: Thank You! done!

> L7 traffic routing is based on L4 with the addition of the Waypoint proxy, which is more complex to handle in Envoy. We can also create HPAs to scale it dynamically.

Reviewer: Not sure if HPA works atm?

> L7 traffic routing is based on L4 with the addition of the Waypoint proxy, which is more complex to handle in Envoy. We can also create HPAs to scale it dynamically.

Reviewer: Saying it based on L4 is confusing.... i would focus on it is optional and provides L7 policy enforcement as needed.

**Overview of the Layer 7 Waypoint proxy**

TODO

**Benefits of using the Layer 7 Waypoint proxy**

TODO

## Pre-requisites & Supported Topologies

Before you begin, make sure that you have the following prerequisites in place:
1. Istio Ambient Mesh installed and configured
2. Ztunnel proxy is installed and running

> 3. Mutual TLS (mTLS) enabled and configured

Reviewer 4: Is this redundant with the tip section as well? Maybe fine

Reviewer: agreed, i would like to see pre-requisites at the beginning of this guide.

> ## Understanding the L7 Waypoint Proxy Default Configuration

Reviewer 2: Better to not call it "L7 Waypoint Proxy" here and everywhere else .. just refer to it as "Waypoint proxy" as is done in the other Ambient docs, blogs etc.  

Try to follow the terminology convention and some of the content from blogs that describe Ambient such as this one ... 
https://istio.io/latest/blog/2023/waypoint-proxy-made-simple/

> ## Understanding the L7 Waypoint Proxy Default Configuration

Reviewer: then we explain to users how to deploy waypoint proxy based on the scope (namespace or service account)

<< Consider breaking this out into bullets for easier reading TODO >>

> The L7 Waypoint Proxy layer in Istio Ambient Mesh is designed to provide a usable configuration out of the box with a fixed feature set that does not require much, or any, custom configuration. Currently, there are no configuration options that need to be set other than the `waypoint` profile setting. Once this profile is used, this in turn sets two internal configuration parameters within the Istio Operator which eventually set the configuration of the L7 Waypoint Proxy.

Reviewer: I'm confused, what is this waypoint profile?

> In the future, there may be some additional limited configurability for L7 Waypoint Proxies. However, for now, the following are all configured with fixed default configurations that are not customizable:

Reviewer 2: Remove this text or section .. this text was appropriate for the ztunnel guide since the ztunnel proxy is designed to work with minimal user configured features and the two internal configuration parameters also refers to the ztunnel proxy. The Waypoint proxy does get configured based on resources such as Kubernetes Gateway API CRDs and Istio/VirtualService etc so this may not be as appropriate as for ztunnel.

* Networking between pods and L7 Waypoint Proxies
* Networking between L7 Waypoint Proxies
* Networking between L7 Waypoint Proxies and sidecar proxies

#### Additional Notes

* The L7 Waypoint Proxy layer can coexist with sidecar proxies in the same cluster. This allows you to use the L7 Waypoint Proxy for services that require L7 functionality and sidecar proxies for services that do not.
* The L7 Waypoint Proxy is implemented using Envoy. This means that you can use all of the features of Envoy in your L7 Waypoint Proxies.
* The L7 Waypoint Proxy is still under development, but it is already a powerful tool for managing L7 traffic in Istio Ambient Mesh.

> For the examples in this guide, we used a deployment of Istio Ambient on a `kind` cluster, although these should apply for any Kubernetes cluster version 1.18.0 or later. Refer to the Getting started guide on how to download the `istioctl` client and how to deploy a `kind` cluster. It would be recommended to have a cluster with more than 1 worker node in order to fully exercise the examples described in this guide. 

Reviewer 4: ```suggestion
For the examples in this guide, we used a deployment of Istio Ambient on a `KinD` cluster, although these should apply for any Kubernetes cluster version 1.18.0 or later. Refer to the Getting started guide on how to download the `istioctl` client and how to deploy a `KinD` cluster. It would be recommended to have a cluster with more than 1 worker node in order to fully exercise the examples described in this guide. 
```

## Functional Overview

> A figure showing an architecture summary of the L7 waypoint proxy.

Reviewer: can we reuse some contents and diagram from the waypoint blog I wrote on istio.io?

TODO

## Install Waypoint Proxy

**Install Gateway CRDs**

> In L7 networking, a waypoint proxy is a lightweight Envoy proxy that runs on each node in the cluster. It is used to implement L7 functionality in Istio Ambient Mesh.

Reviewer 4: ```suggestion
In L7 networking, a waypoint proxy is a lightweight Envoy proxy that can be configured for your entire namespace or for a service account. It is used to implement L7 functionality in Istio Ambient Mesh.
```

Waypoints aren't daemonsets so they don't necessarily run on every node.

> Waypoint proxies are dependent on Gateway API CRDs to provide features such as traffic routing and service discovery.

Reviewer 4: ```suggestion
The reference implementation of a waypoint proxy is managed by the Kubernetes Gateway API `istio-waypoint` GatewayClass.
```

> 1. Install Kubernetes Gateway CRDs, which don’t come installed by default on most Kubernetes clusters:

Reviewer 2: Maybe use the term "Kubernetes Gateway API CRDs" to be a bit more explicit

    {{< text bash >}}
    $ kubectl get crd gateways.gateway.networking.k8s.io &> /dev/null || \
      { kubectl kustomize "github.com/kubernetes-sigs/gateway-api/config/crd/experimental?ref={{< k8s_gateway_api_version >}}" | kubectl apply -f -; }
    {{< /text >}}

> 

Reviewer: not sure if we need this, i'd expect the pre-requisite cover this.

The core functionality of the waypoint L7 traffic management is identical to sidecar mode, hence to add more features refer to this link: https://istio.io/latest/docs/reference/config/networking/virtual-service/

FOR EACH SECTION:

### Creating a Virtual Service

To create a virtual service for the Bookinfo application, you can use the following YAML manifest:

{{< text bash >}}
$ yaml file
{{< /text >}}

This virtual service will route all traffic to the `bookinfo-v1` service.

### Deploying the Virtual Service

To deploy the virtual service, you can use the following command:

{{< text bash >}}
$ command
{{< /text >}}

### Verifying the Waypoint Proxy Configuration is working

Once the virtual service is deployed, you can verify the waypoint proxy configuration by running the following command:

{{< text bash >}}
$ command
{{< /text >}}

This will output the configuration of the waypoint proxy, including the virtual service that is mapped to it.

### Configuring Virtual Services

> You can configure both L4 and L7 virtual services for waypoint proxies. If you want to do a Virtual Service with TCP that is effectively a L4 virtual services. This is used to route traffic to services based on port number. L7 virtual services are used to route traffic to services based on more complex criteria, such as HTTP method and path. In a Virtual service you can have only TCP, or only HTTP or Both.

Reviewer 2: ```suggestion
You can configure both L4 and L7 virtual services for waypoint proxies. If you want to do a Virtual Service with TCP that is effectively a L4 virtual services. This is used to route traffic to services based on port number. L7 virtual services are used to route traffic to services based on more complex criteria, such as HTTP method and path. In a Virtual service you can have only TCP, or only HTTP or Both.
```
A VirtualService defines a set of traffic routing rules to apply when a host is addressed. The functional behavior of VirtualService is identical in Istio ambient mode as in sidecar mode.  Refer to the Istio Traffic Management and VirtualService guides for the exact definition and details. Each routing rule defines matching criteria for traffic of a specific protocol (which currently can be either TCP or HTTP).

<< Considered merging L4 (TCP) and L7 (HTTP) virtual service >>

### Example

The following YAML manifest shows an example of an L7 virtual service:

{{< text bash >}}
$ yaml file
{{< /text >}}

This virtual service will route traffic to the `bookinfo-v1` service for requests to the `/productpage` path and traffic to the `bookinfo-v2` service for requests to the `/reviews` path.

### Verifying the Virtual Service

Once the virtual service is set up, the HTTP route is mapped to the waypoint configuration. This means that all traffic that matches the virtual service's hosts and HTTP routes will be routed to the waypoint proxy.

> You can verify this by using the `istioctl proxy-config` command to dump the waypoint configuration. This will show you the virtual services that are mapped to the waypoint proxy.

Reviewer 3: "You can verify this by using the `istioctl proxy-config` command to dump the waypoint configuration. This will show you the virtual services that are mapped to the waypoint proxy."

Might want to change to something like:
In classic mode you can use `istioctl proxy-config cmds` to dump the envoy configuration. While `istioctl proxy-config cmds` will still work in ambient there are some differences because the envoy proxy is no longer configured for every sidecar.    

Would then follow the section about the istioctl proxy-config cmds found here https://istio.io/latest/blog/2023/waypoint-proxy-made-simple/

* **Use the Istio telemetry dashboards:** The Istio telemetry dashboards provide a graphical view of the traffic flowing through your Istio mesh. You can use these dashboards to monitor the traffic flowing to your waypoint proxies and to identify any problems with the virtual service mapping.
* **Use the Istio `istioctl` command:** You can use the Istio `istioctl` command to view the waypoint configuration. This configuration includes the virtual services that are mapped to the waypoint.
* **Use Prometheus and Grafana:** You can use Prometheus and Grafana to collect and visualize metrics from your Istio mesh. You can use these tools to monitor the metrics associated with the waypoint proxy configuration and the virtual service mapping.

By monitoring the waypoint proxy and the virtual service mapping, you can ensure that your Istio mesh is operating as expected.

## L7 Authorization Policy
TODO

## Monitoring and Telemetry with L7 Waypoint Proxy
TODO

## How to use the Waypoint proxy for hairpinning
TODO

## Co-existence of Ambient/ L7 with Side car proxies
How to use the Waypoint proxy with sidecar proxies
TODO

> 

Reviewer 3: do the following examples have anything unique to waypoint or are they just additional examples?  I think they may just work the same way so we don't really need to include them

owner: istio/wg-networking-maintainers
test: no
---

{{< warning >}}
Ambient is currently in [alpha status](/docs/releases/feature-stages/#feature-phase-definitions).

> Please **do not run ambient in production** and be sure to thoroughly review the [feature phase definitions](/docs/releases/feature-stages/#feature-phase-definitions) before use. In particular, there are known performance, stability, and security issues in the `alpha` release. There are also functional caveats some of which are listed in the [Caveats section](#caveats) of this guide. There are also planned breaking changes, including some that will prevent upgrades. These are all limitations that will be addressed before graduation to `beta`. The current version of this guide is meant to assist early deployments and testing of the alpha version of `ambient`. The guide will continue to get updated as `ambient` itself evolves from alpha to beta status and beyond.

Reviewer 5: ```suggestion
Please **do not run ambient in production** and be sure to thoroughly review the [feature phase definitions](/docs/releases/feature-stages/#feature-phase-definitions) before use. In particular, there are known performance, stability, and security issues in the `alpha` release. There are also functional caveats some of which are listed in the [Caveats section](#caveats) of this guide. There are also planned breaking changes, including some that will prevent upgrades. These are all limitations that will be addressed before graduation to `beta`. The current version of this guide is meant to assist early deployments and testing of the alpha version of ambient. The guide will continue to get updated as ambient itself evolves from alpha to beta status and beyond.
```

{{< /warning >}}

## Introduction

> This guide provides instructions on how to set up and use the Waypoint proxy layer in Istio Ambient Mesh.

Reviewer 4: +1 to standardizing
+1 to lowercase for the standard

> Istio Ambient Mesh is a new way to deploy and manage microservices. In Ambient Mesh, workloads are no longer required to run sidecar proxies to participate in the service mesh. Ambient splits Istio’s functionality into two distinct layers, a secure overlay layer and a Layer 7 processing layer.

Reviewer 6: Is ambient without mesh ok? I'm not sure.
```suggestion
Istio ambient mesh is a new way to deploy and manage microservices. In ambient mesh, workloads are no longer required to run sidecar proxies to participate in the service mesh. Ambient splits Istio’s functionality into two distinct layers, a secure overlay layer and a Layer 7 processing layer.
```

Ztunnel is used to handle L3 and L4 networking functions, such as mTLS authentication and L4 authorization. The waypoint proxy is an optional component that is Envoy-based and is responsible for L7 service mesh functionalities. For workloads that require L7 networking features, such as HTTP load balancing and fault injection, a waypoint proxy must be deployed.  It can also enforce L7 policies and collect L7 metrics.

{{<tip>}}

<!-- Pre-requisites & Supported Topologies -->

Before you begin, make sure that you have already read the [Ztunnel Networking sub-guide](../ztunnel/). This guide assumes that you have the following prerequisites in place:

> 1. Istio Ambient Mesh installed and configured

Reviewer 6: ```suggestion
1. Istio ambient mesh installed and configured
```

2. Ztunnel proxy is installed and running

{{</tip>}}

> This guide describes the functionality and usage of the waypoint proxy and L7 networking functions using Istio Ambient Mesh. We use a sample user journey to describe these functions hence it would be useful to go through this guide in sequence. However we provide links to the sections below in case the reader would like to jump to the appropriate section.

Reviewer 6: ```suggestion
This guide describes the functionality and usage of the waypoint proxy and L7 networking functions using Istio ambient mesh. We use a sample user journey to describe these functions hence it would be useful to go through this guide in sequence. However we provide links to the sections below in case the reader would like to jump to the appropriate section.
```

* [Introduction](#introduction)
* [Deciding if you need A Waypoint Proxy](#deciding-if-you-need-a-waypoint-proxy)
* [Current Challenges](#current-challenges)
<!-- * [Differences between Sidecar Mode and Ambient Mode for Waypoint Proxy](#differences) -->
* [Deciding the scope of your Waypoint Proxy](#Deciding-the-scope-of-your-Waypoint-Proxy)
* [Functional Overview](#functional-overview)
* [Deploying an Application](#deploying-an-application)
* [Configuring Waypoint proxy](#configuring-waypoint-proxy)
* [Monitoring the Waypoint Proxy & L7 Networking](#monitoring-the-waypoint-proxy--l7-networking)
* [L7 Fault Injection](#l7-fault-injection)

> * [L7 Observability](#l7-observability)

Reviewer 6: ```suggestion
* [L7 observability](#l7-observability)
```

> * [L7 Authorization Policy](#l7-authorization-policy)

Reviewer 6: ```suggestion
* [L7 authorization policy](#l7-authorization-policy)
```

> * [Co-existence of Ambient/ L7 with Side car proxies](#Co-existence-of-Ambient/-L7-with-Side-car-proxies)

Reviewer 6: Not sure about ambient. We haven't capitalized it in the past.
```suggestion
* [Co-existence of ambient/ L7 with Side car proxies](#Co-existence-of-Ambient/-L7-with-side-car-proxies)
```

> * [Co-existence of ambient/ L7 with Side car proxies](#Co-existence-of-ambient/-L7-with-side-car-proxies)

Reviewer 7: ```suggestion
* [Co-existence of ambient/ L7 with Sidecar proxies](#Co-existence-of-ambient/-L7-with-side-car-proxies)
```

> * [Control Traffic towards Waypoint Proxy](##control-traffic-towards-waypoint-proxy)

Reviewer 6: ```suggestion
* [Control traffic towards waypoint proxy](##control-traffic-towards-waypoint-proxy)
```

> * [Remove Waypoint proxy layer](#remove-waypoint-proxy-layer)

Reviewer 6: ```suggestion
* [Remove waypoint proxy layer](#remove-waypoint-proxy-layer)
```

> ## Deciding if you need A Waypoint proxy

Reviewer 6: ```suggestion
## Deciding if you need a waypoint proxy
```

It's possible that the features offered by the secure overlay doesn’t meet your requirements. For instance, you need a rich Layer 7 authorization policy that sets up access based on a certain method and path. Alternatively you may like to conduct a canary test on the updated version of your service or introduce a new version without affecting current traffic. Or, you would like to receive metrics, HTTP access logs, and distributed tracing for some of your services. In order to accomplish these common cases, we'll go over how you can choose to enforce L7 processing with ambient mesh in this section.

### Benefits of using the waypoint proxy and L7 networking features

In summary, the waypoint proxy approach for the L7 processing layer offers the following four main advantages:

> * Traffic Resiliency - Timeout and retry, circuit breaking

Reviewer 6: We have a lower case management in a lower item.
```suggestion
* Traffic resiliency - Timeout and retry, circuit breaking
```

* L7 fault injection: You need to simulate faults in your microservices architecture such as delays, errors, and circuit breaks to test its resilience and prepare for real-world failures.
* Rate limiting: You need to protect workloads against denial-of-service attacks and improve performance.
* L7 observability: You need to collect metrics and traces from your microservices architecture to monitor its performance and troubleshoot problems.
* Rich Authz Policy: Enforce granular authorization control and manage authorization policies centrally as well as enhanced Security

### Getting started with the waypoint proxy and L7 networking features

To get started with the waypoint proxy and L7 networking features, you will need to deploy a waypoint proxy for each workload that requires L7 networking. You can do this using the Kubernetes Gateway resource. Once the waypoint proxy is deployed, you can configure L7 policies using the VirtualService, DestinationRule, and ServiceEntry resources.

This guide will provide more detailed instructions on how to deploy and configure the waypoint proxy and L7 networking features.

> ## Current Challenges

Reviewer 6: ```suggestion
## Current challenges
```

1. **`PeerAuthentication` Limitations:** As of now, not all components (i.e. waypoint proxies), support the `PeerAuthentication` resource in Istio ambient mode. Hence it is recommended to use the `STRICT` mTLS mode currently, this caveat shall be addressed as the feature moves toward beta status.

1. **No HTTP/3 support:** Waypoint supports HTTP/2, but there are some limitations with HTTP/3 support when using HBONE (HTTP-Based Overlay Network Encapsulation). HBONE is currently limited to HTTP/2 transport, so while waypoint can handle HTTP/2 traffic, it cannot yet fully support HTTP/3.

Despite these caveats, waypoint is a powerful tool for enabling L7 networking and services for Istio ambient workloads. It is a good choice for users who are looking for a way to run microservices-based applications in ambient mode.

### Environment used for this guide

> For the environment used in this guide, refer to the [Installation user](../install/) guide or [L4 Networking & mTLS with Ztunnel](../ztunnel/) guide on installing Istio in ambient mode on a Kubernetes cluster.

Reviewer 5: ```suggestion
For the environment used in this guide, refer to the [Installation user](../../install/) guide or [L4 Networking & mTLS with Ztunnel](../ztunnel/) guide on installing Istio in ambient mode on a Kubernetes cluster.
```

## Deciding the scope of your Waypoint proxy

Waypoint proxies can be deployed at the namespace(default) or service account level. The scope you choose depends on your specific needs and requirements.

### Namespace-level scope

Deploying Waypoint proxies at the namespace level provides a number of benefits, including:
- Simplified policy management: Policies are enforced at the namespace level, so you only need to define them once for all workloads in the namespace.
- Improved performance: Waypoint proxies can cache routing and policy information, which can improve performance for workloads in the namespace.
- Increased security: Waypoint proxies can enforce authorization policies at the namespace level, which can help to protect your workloads from unauthorized access.

> However, deploying Waypoint proxies at the namespace level also has some drawbacks, including:

Reviewer 6: ```suggestion
However, deploying waypoint proxies at the namespace level also has some drawbacks, including:
```

- Limited granularity: You cannot apply different policies to different workloads in the same namespace.

> - Increased resource consumption: Each namespace will require its own Waypoint proxy, which can consume more resources.

Reviewer 6: ```suggestion
- Increased resource consumption: Each namespace will require its own waypoint proxy, which can consume more resources.
```

### Service account-level scope

> Deploying Waypoint proxies at the service account level provides a number of benefits, including:

Reviewer 6: ```suggestion
Deploying waypoint proxies at the service account level provides a number of benefits, including:
```

- Increased granularity: You can apply different policies to different workloads based on their service account.

> - Reduced resource consumption: You only need to deploy a Waypoint proxy for each service account that has workloads that require L7 routing or policy enforcement.

Reviewer 6: ```suggestion
- Reduced resource consumption: You only need to deploy a waypoint proxy for each service account that has workloads that require L7 routing or policy enforcement.
```

> However, deploying Waypoint proxies at the service account level also has some drawbacks, including:

Reviewer 6: ```suggestion
However, deploying waypoint proxies at the service account level also has some drawbacks, including:
```

* Increased complexity: Managing policies at the service account level can be more complex, especially if you have a large number of service accounts.
* Reduced caching: Waypoint proxies cannot cache routing and policy information at the service account level, which can reduce performance.

> ### How to choose the right scope for your Waypoint proxies

Reviewer 6: ```suggestion
### How to choose the right scope for your waypoint proxies
```

> The best scope for your Waypoint proxies will depend on your specific needs and requirements. If you have a simple application with a small number of workloads, then namespace-level scope may be a good choice. However, if you have a more complex application with a large number of workloads or if you need to apply different policies to different workloads, then service account-level scope may be a better choice.

Reviewer 6: ```suggestion
The best scope for your waypoint proxies will depend on your specific needs and requirements. If you have a simple application with a small number of workloads, then namespace-level scope may be a good choice. However, if you have a more complex application with a large number of workloads or if you need to apply different policies to different workloads, then service account-level scope may be a better choice.
```

Here are some factors to consider when choosing the scope for your Waypoint proxies:
* The number of workloads in your application
* The complexity of your application
* The need to apply different policies to different workloads
* The performance requirements of your application
* The resource requirements of your application

> ## Functional Overview

Reviewer 6: ```suggestion
## Functional overview
```

> The functional behavior of the waypoint proxy is dynamically configured by Istio to serve your applications configurations. This section takes a brief look at these functional aspects - detailed description of the internal design of the waypoint proxy is out of scope for this guide. The detailed functional overview from the Secure Overlay Networking was already discussed in the [Ztunnel L4 Networking Guide](https://deploy-preview-13635--preliminary-istio.netlify.app/latest/docs/ops/ambient/usage/ztunnel/#functionaloverview) hence this section only focuses on functionalities and features that waypoint proxy provides.

Reviewer 7: ```suggestion
The functional behavior of the waypoint proxy is dynamically configured by Istio to serve your application's configurations. This section takes a brief look at these functional aspects - detailed description of the internal design of the waypoint proxy is out of scope for this guide. The detailed functional overview from the Secure Overlay Networking was already discussed in the [Ztunnel L4 Networking Guide](https://deploy-preview-13635--preliminary-istio.netlify.app/latest/docs/ops/ambient/usage/ztunnel/#functionaloverview) hence this section only focuses on functionalities and features that waypoint proxy provides.
```

{{< image width="100%"
link="waypoint-architecture.png"
caption="Waypoint architecture"
>}}

What is unique about the waypoint proxy is that it runs either per-namespace (default) or per-service account. By running outside of the application pod, a waypoint proxy can install, upgrade, and scale independently from the application pod, providing a centralized approach to managing L7 traffic and enforcing policies as well as reduce operational costs.

> Upon deployment of a gateway resource using the `kubectl apply` command, Istio's control plane, IstioD, assumes the role of the waypoint controller. Recognizing the gateway resource with the "istio.io/waypoint" gateway class name, istiod automatically deploys the waypoint proxy based on the specified configuration in the gateway resource.

Reviewer 7: I think the docs usually have istiod: https://github.com/istio/istio.io/blob/cf7f060c5c0f50eb187685bdb6638b139f147e65/content/en/docs/reference/glossary/istiod.md?plain=1#L6 

```suggestion
Upon deployment of a gateway resource using the `kubectl apply` command, Istio's control plane, istiod, assumes the role of the waypoint controller. Recognizing the gateway resource with the "istio.io/waypoint" gateway class name, istiod automatically deploys the waypoint proxy based on the specified configuration in the gateway resource.
```

caption="The waypoint proxy is deployed per service account/ workload identity and can be thought of as a 'gateway per workload'"
>}}

The deployment of waypoint proxies can be handled by namespace owners, platform operators, or automated systems. Once a waypoint proxy is deployed, and a corresponding L7 policy is configured for a destination represented by the waypoint proxy, the secure overlay layer ensures that connection is routed to the correct L7 waypoint proxy for processing and policy enforcement as shown here -

{{< image width="100%"
link="waypoint-traffic-flow.svg"
caption="Traffic will flow through L7 waypoint proxies when there are L7 policies that need to be enforced for a particular service"
>}}

Tenancy for Layer 7 capabilities in the Istio ambient mesh is similar to the sidecar deployment model. L7 capabilities are not shared across multiple identities within a single L7 proxy. Each application has its own dedicated waypoint proxy, ensuring isolation of configuration and extensions (plug-ins, extensions, etc.) specific to individual workloads. This isolation prevents interference between workloads and facilitates independent management of L7 configurations.

> Functionally, the Waypoint proxy resembles the sidecar proxy but operates independently of application pods. It has its own CA client and XDS client, enabling secure communication with istioD. To obtain its identity certificate, the Waypoint proxy establishes a secure connection with istioD, requesting certification. Upon validating the presented token, istioD signs the Waypoint proxy's certificate, granting it access to the Istio control plane.

Reviewer 6: ```suggestion
Functionally, the waypoint proxy resembles the sidecar proxy but operates independently of application pods. It has its own CA client and XDS client, enabling secure communication with istioD. To obtain its identity certificate, the waypoint proxy establishes a secure connection with istioD, requesting certification. Upon validating the presented token, istioD signs the waypoint proxy's certificate, granting it access to the Istio control plane.
```

> Functionally, the waypoint proxy resembles the sidecar proxy but operates independently of application pods. It has its own CA client and XDS client, enabling secure communication with istioD. To obtain its identity certificate, the Waypoint proxy establishes a secure connection with istioD, requesting certification. Upon validating the presented token, istioD signs the Waypoint proxy's certificate, granting it access to the Istio control plane.

Reviewer 7: ```suggestion
Functionally, the waypoint proxy resembles the sidecar proxy but operates independently of application pods. It has its own CA client and XDS client, enabling secure communication with istiod. To obtain its identity certificate, the Waypoint proxy establishes a secure connection with istiod, requesting certification. Upon validating the presented token, istiod signs the Waypoint proxy's certificate, granting it access to the Istio control plane.
```

> Subsequently, the Waypoint proxy initiates communication with istioD, requesting XDS configuration to govern its operation. This configuration defines the L7 routing rules, policy enforcement mechanisms, and other parameters essential for managing L7 traffic.

Reviewer 6: ```suggestion
Subsequently, the waypoint proxy initiates communication with istioD, requesting XDS configuration to govern its operation. This configuration defines the L7 routing rules, policy enforcement mechanisms, and other parameters essential for managing L7 traffic.
```

> In essence, the Waypoint proxy serves as an L7 traffic management hub, decoupled from application pods and centrally managed by istioD. This architecture simplifies L7 configuration and policy enforcement, enabling efficient and scalable L7 services within Istio Ambient deployments as shown in the figure -

Reviewer 6: ```suggestion
In essence, the waypoint proxy serves as an L7 traffic management hub, decoupled from application pods and centrally managed by istioD. This architecture simplifies L7 configuration and policy enforcement, enabling efficient and scalable L7 services within Istio ambient deployments as shown in the figure -
```

{{< image width="100%"
link="waypoint-architecture-deep-dive.svg"
caption="Waypoint Architecture Deep Dive"
>}}

> ### Destination Only Waypoint

Reviewer 6: ```suggestion
### Destination only waypoint
```

> In contrast to traditional sidecar proxies, which reside alongside application pods, Waypoint proxies operate solely on the server-side, acting as reverse proxies for L7 traffic. This approach streamlines L7 traffic management by centralizing policy enforcement to the destination workload's namespace or service account.

Reviewer 6: ```suggestion
In contrast to traditional sidecar proxies, which reside alongside application pods, waypoint proxies operate solely on the server-side, acting as reverse proxies for L7 traffic. This approach streamlines L7 traffic management by centralizing policy enforcement to the destination workload's namespace or service account.
```

> When a request originates from an application pod, it bypasses the client-side Waypoint proxy and directly reaches the server-side Waypoint proxy associated with the destination workload's namespace or service account. Istio enforces that all traffic coming into the namespace goes through the waypoint, which then enforces all policies for that namespace. Because of this, each waypoint only needs to know about configuration for its own namespace. Thus Waypoint proxy assumes responsibility for enforcing all L7 policies and routing rules applicable to the destination workload.

Reviewer 6: ```suggestion
When a request originates from an application pod, it bypasses the client-side waypoint proxy and directly reaches the server-side waypoint proxy associated with the destination workload's namespace or service account. Istio enforces that all traffic coming into the namespace goes through the waypoint, which then enforces all policies for that namespace. Because of this, each waypoint only needs to know about configuration for its own namespace. Thus waypoint proxy assumes responsibility for enforcing all L7 policies and routing rules applicable to the destination workload.
```

{{< image width="100%"
link="destinationonly.svg"
caption="Waypoint Proxies"
>}}

> Destination-only Waypoint simplifies the configuration process by eliminating the need for sidecar proxies and `exportTo` configurations. Waypoint proxies only need to be aware of the endpoints, pods, and workloads within their respective namespaces or service accounts. This streamlined approach reduces the complexity of L7 management and enables a more efficient use of resources.

Reviewer 6: ```suggestion
Destination-only waypoint simplifies the configuration process by eliminating the need for sidecar proxies and `exportTo` configurations. Waypoint proxies only need to be aware of the endpoints, pods, and workloads within their respective namespaces or service accounts. This streamlined approach reduces the complexity of L7 management and enables a more efficient use of resources.
```

{{< image width="100%"
link="destination-only-waypoint.svg"
caption="Waypoint Proxies"
>}}

* **Policy Enforcement**: In traditional Istio deployments, both source-side and destination-side policies were employed, which often led to confusion for users regarding policy enforcement and troubleshooting. Destination-only Waypoint simplifies this process by enforcing all policies exclusively at the destination workload's namespace or service account. This centralized approach eliminates the need to track policies across multiple locations, making it easier to understand, manage, and troubleshoot L7 security configurations.

{{< image width="100%"
link="policies-enforced (1).svg"
caption="Policy Enforced on Destination Waypoint"
>}}

* **Mixed Environment**: In a mixed environment where clients may reside inside or outside the Istio mesh, destination-only Waypoint ensures consistent policy enforcement regardless of the client's location. Since all policies are applied at the destination workload, users can be confident that security measures are consistently applied to all incoming traffic.

{{< image width="100%"
link="mixed-environment.svg"
caption="Waypoint Proxies"
>}}

> ### Handling Destinations without Waypoint Proxies

Reviewer 6: ```suggestion
### Handling destinations without waypoint proxies
```

> While destination-only Waypoint offers centralized policy enforcement and simplified configuration, there may be instances where the destination workload doesn't have a waypoint proxy deployed. This could arise when connecting to external services beyond the control of the Istio mesh.

Reviewer 6: ```suggestion
While destination-only waypoint offers centralized policy enforcement and simplified configuration, there may be instances where the destination workload doesn't have a waypoint proxy deployed. This could arise when connecting to external services beyond the control of the Istio mesh.
```

To address this scenario, the Istio community is actively developing mechanisms to route traffic to the egress gateway and enable policy enforcement for destinations without waypoint proxies. This functionality will allow users to configure resilience-enhancing policies, such as timeouts, for external services.

Please stay tuned for future blog posts and documentation updates that will provide detailed information on this evolving feature.

> ## Deploying an Application

Reviewer 6: ```suggestion
## Deploying an application
```

When someone with Istio admin privileges sets up Istio mesh, it becomes available for all users in specific namespaces. The examples below shows how Istio can be used transparently once it's successfully deployed in ambient mode and the namespaces are annotated accordingly.

### Basic application deployment without ambient

> This section is Under Construction.

Reviewer 7: This would a be a great thing to show- you can deploy the example apps, _then_ install Istio in ambient mode without having to restart the applications. You can probably use the sleep/notsleep example apps from Lin's getting started guide: https://istio.io/latest/blog/2022/get-started-ambient/#deploy-your-applications. 

If you wanted this section to show how you can have sidecars injected in an ambient-labeled namespace: 

```
kubectl label namespace default istio.io/dataplane-mode=ambient --overwrite=true
kubectl apply -f sleep/sleep.yaml -n default

# notsleep is in sidecar mode
kubectl apply -f <(istioctl kube-inject -f sleep/notsleep.yaml) -n default
```

In particular, there are known performance, stability, and security issues in the `alpha` release. There are also functional caveats some of which are listed in the [Caveats section](#caveats) of this guide. There are also planned breaking changes, including some that will prevent upgrades. These are all limitations that will be addressed before graduation to `beta`. The current version of this guide is meant to assist early deployments and testing of the alpha version of `ambient`. The guide will continue to get updated as `ambient` itself evolves from alpha to beta status and beyond. 
{{< /warning >}}

## Introduction

This guide provides instructions on how to set up and use the Waypoint proxy layer in Istio Ambient Mesh.

Istio Ambient Mesh is a new way to deploy and manage microservices. In Ambient Mesh, workloads are no longer required to run sidecar proxies to participate in the service mesh. Ambient splits Istio’s functionality into two distinct layers, a secure overlay layer and a Layer 7 processing layer.

> Ztunnel proxy is used to handle L3 and L4 networking functions, such as mTLS authentication and L4 authorization. For workloads that require L7 networking features, such as HTTP load balancing and fault injection, a waypoint proxy can be deployed. The waypoint proxy is an optional component that is Envoy-based and is responsible for terminating workload HTTP traffic and parsing workload HTTP headers. They also enforce L7 policies and collect L7 metrics.

Reviewer: ```suggestion
Ztunnel is used to handle L3 and L4 networking functions, such as mTLS authentication and L4 authorization. The waypoint proxy is an optional component that is Envoy-based and is responsible for terminating workload HTTP traffic and parsing workload HTTP headers. For workloads that require L7 networking features, such as HTTP load balancing and fault injection, a waypoint proxy must be deployed.  It can also enforce L7 policies and collect L7 metrics.
```

{{<tip>}}

<!-- Pre-requisites & Supported Topologies -->

Before you begin, make sure that you have already read the [Ztunnel Networking sub-guide](../ztunnel/). This guide assumes that you have the following prerequisites in place:
1. Istio Ambient Mesh installed and configured
2. Ztunnel proxy is installed and running

> 3. Mutual TLS (mTLS) enabled and configured

Reviewer: you mean enforced?

mtls is enabled by default.

* [Co-existence of Ambient/ L7 with Side car proxies](#Co-existence-of-Ambient/-L7-with-Side-car-proxies)
* [Control Traffic towards Waypoint Proxy](##control-traffic-towards-waypoint-proxy)
* [Remove Waypoint proxy layer](#remove-waypoint-proxy-layer)

## Deciding if you need A Waypoint proxy

It's possible that the features offered by the secure overlay doesn’t meet your requirements. For instance, you need a rich Layer 7 authorization policy that sets up access based on a certain method and path. Alternatively you may like to conduct a canary test on the updated version of your service or introduce a new version without affecting current traffic. Or, you would like to receive metrics, HTTP access logs, and distributed tracing for some of your services. In order to accomplish these common cases, we'll go over how you can choose to enforce L7 processing with ambient mesh in this section.

### Benefits of using the waypoint proxy and L7 networking features

In summary, the waypoint proxy approach for the L7 processing layer offers the following three main advantages:

- Security - Rich L7 authorization policy
- Observability - HTTP metrics, access logs, and tracing

> - Traffic management - Dark launch, canary test

Reviewer: also traffic resiliency.

The waypoint proxy and L7 networking features provide a number of benefits, including:

- Improved performance and scalability: Waypoint proxies are designed to be lightweight and efficient, which can improve the performance and scalability of your microservices architecture.
- Increased flexibility: The waypoint proxy allows you to implement a wide range of L7 networking features, such as HTTP load balancing, fault injection, and observability.

> - Simplified operations: By deploying a waypoint proxy, you can simplify the operation of your microservices architecture by reducing the number of components that need to be managed.

Reviewer: I suggest remove this section as none of these benefits are convincing comparing with sidecars.

### When to use the waypoint proxy and L7 networking features

You should consider using the waypoint proxy and L7 networking features if your microservices architecture requires any of the following:

- L7 load balancing and routing: You need to distribute traffic across multiple instances of a workload based on factors such as request path, header values, or cookies.
Waypoint provides a variety of L7 load balancing and routing algorithms, including round robin, weighted round robin, and least connections. It also supports path-based routing and other advanced routing rules.
- L7 fault injection: You need to simulate faults in your microservices architecture such as delays, errors, and circuit breaks to test its resilience and prepare for real-world failures.

> - Rate limiting: You need to protect workloads against denial-of-service attacks and improve performance.

Reviewer: I would put these into a traffic resiliency section, also add timeout and retries.

- L7 observability: You need to collect metrics and traces from your microservices architecture to monitor its performance and troubleshoot problems.

> 

Reviewer: also add rich Authz policy here.

### Getting started with the waypoint proxy and L7 networking features

To get started with the waypoint proxy and L7 networking features, you will need to deploy a waypoint proxy for each workload that requires L7 networking. You can do this using the Kubernetes Gateway resource. Once the waypoint proxy is deployed, you can configure L7 policies using the VirtualService, DestinationRule, and ServiceEntry resources.

This guide will provide more detailed instructions on how to deploy and configure the waypoint proxy and L7 networking features.

## Current Challenges

> Unlike Ztunnel proxies, Waypoint proxies are not automatically installed with Istio ambient mesh. Waypoint proxies are deployed declaratively using Kubernetes Gateway resources or the helpful `istioctl` command. The minimum Istio version required for Istio ambient mode is `1.18.0`. In general Istio in ambient mode supports the existing Istio APIs that are supported in sidecar proxy mode. Since the ambient functionality is currently at an alpha release level, the following is a list of feature restrictions or caveats in the current release of Istio's ambient functionality (as of the `1.19.0` release). These restrictions are expected to be addressed/removed in future software releases as ambient graduates to beta and eventually General Availability.

Reviewer: latest releaase is 1.20.0 now :)

1. **Kubernetes only:** Istio in ambient mode is currently limited to deployment on Kubernetes clusters, excluding non-Kubernetes endpoints like virtual machines.

2. **Single Cluster Support:** Multi-cluster deployments are not supported in Istio ambient mode; only single-cluster configurations are currently viable.

3. **K8s CNI Restrictions:** Istio in ambient mode does not currently work with every Kubernetes CNI implementation. Additionally, with some plugins, certain CNI functions (in particular Kubernetes `NetworkPolicy` and Kubernetes Service Load balancing features) may get transparently bypassed in the presence of Istio ambient mode. The exact set of supported CNI plugins as well as any CNI feature caveats are currently under test and will be formally documented as Istio's ambient mode approaches the beta release.

> 4. **TCP/IPv4 only:** In the current release, TCP over IPv4 is the only protocol supported for transport on an Istio secure overlay tunnel (this includes protocols such as HTTP that run between application layer endpoints on top of the TCP/ IPv4 connection).

Reviewer: suggest moving these to L4 section, none of this is unique to waypoint.

cc @srampal

5. **No Dynamic switching to Ambient mode:** Enabling ambient mode is only possible during the deployment of a new Istio mesh control plane using an ambient profile or helm configuration. An existing Istio mesh deployed using a pre-ambient profile for instance can not be dynamically switched to also enable ambient mode operation.

6. **Restrictions with Istio `PeerAuthentication`:** as of the time of writing, the `PeerAuthentication` resource is not supported by all components (i.e. waypoint proxies) in Istio ambient mode. Hence it is recommended to only use the `STRICT` mTLS mode currently. Like many of the other alpha stage caveats, this shall be addressed as the feature moves toward beta status.

6. **`PeerAuthentication` Limitations:** As of now, not all components (i.e. Waypoint proxies), support the `PeerAuthentication` resource in Istio Ambient mode. Hence it is recommended to use the `STRICT` mTLS mode currently, this caveat shall be addressed as the feature moves toward beta status.

> 

Reviewer: dups?  

I think this point is valid.  cc @keithmattix if there is any objections.

Reviewer 8: Yes this is correct (and it's duplicated)

7. **istioctl CLI Gaps:** Minor functional gaps may exist in the Istio CLI's output displays related to ambient mode. These will be addressed as the feature matures.

In addition to this general caveats, there are also some specific caveats to be aware of when using Waypoint with certain protocols:

- Waypoint only supports Ambient workloads. It does not support sidecar proxy workloads. 
- HTTP: Waypoint does not support all HTTP features, such as HTTP/2 and chunked encoding.
- gRPC: Waypoint does not support all gRPC features, such as HTTP/2 transport and protocol multiplexing.

> - WebSocket: Waypoint does not support WebSocket.

Reviewer: I am not sure if all of these are correct. waypoint can work with sidecars and should be able to support HTTP/2.

If there is limitation, it would be how we implement HBONE where we don't support HTTP/3 yet.

cc @howardjohn

Despite these caveats, Waypoint is a powerful tool for enabling L7 networking and services for Istio Ambient workloads. It is a good choice for users who are looking for a way to run microservices-based applications in Ambient mode.

Here is a table summarizing the caveats of Waypoint:

| Caveat | Description |
| ------------- | ------------- |
| Maturity | Waypoint is still under development |

> | Features | Waypoint only supports L7 load balancing and routing |

Reviewer: waypoint supports more than that.  resiliency and L7 rich authz.

|Integration | Waypoint is not yet fully integrated with the Istio control plane | Support | Waypoint only supports Ambient workloads | HTTP | Waypoint does not support all HTTP features | 

> gRPC | Waypoint does not support all gRPC features | Websocket | Waypoint does not support WebSocket |

Reviewer: pending discussion from https://github.com/istio/istio.io/pull/13954/files#r1414064107, this table needs to be updated

In addition to these caveats, it is also important to note that Waypoint is a new feature, and it is not yet as mature as Istio's sidecar proxy. As a result, users may experience some performance or stability issues when using Waypoint. However, the Ambient mesh team is actively working to address these issues, and they are committed to making Waypoint a production-ready feature.

> Overall, Waypoint is a powerful tool for enabling L7 networking and services for Istio Ambient workloads. However, users should be aware of the caveats and limitations listed above before deploying Waypoint in production.

Reviewer: suggest removing these as felt repetitive.

### Environment used for this guide

> For the examples in this guide, we used a deployment of Istio version `1.19.0` on a `kinD` cluster of version `0.20.0` running Kubernetes version `1.27.3`. However these should also work on any Kubernetes cluster at version `1.24.0` or later and Istio version `1.18.0` or later. It would be recommended to have a cluster with more than 1 worker node in order to fully exercise the examples described in this guide. Refer to the [Installation user](https://deploy-preview-13635--preliminary-istio.netlify.app/latest/docs/ops/ambient/usage/install/) guide or [Getting started guide](https://deploy-preview-13635--preliminary-istio.netlify.app/latest/docs/ops/ambient/getting-started/) on installing Istio in ambient mode on a Kubernetes cluster.

Reviewer: Use relative pathes, for example /docs/ops/ambient/usage/install/

Reviewer: I'd assume just pointing to env from L4 or get starte guide. no need to cover env here again.

## Deciding the scope of your Waypoint proxy

> Waypoint proxies can be deployed at the namespace or service account level. The scope you choose depends on your specific needs and requirements.

Reviewer: need to mention namespace is the default scope.

### Namespace-level scope

Deploying Waypoint proxies at the namespace level provides a number of benefits, including:
- Simplified policy management: Policies are enforced at the namespace level, so you only need to define them once for all workloads in the namespace.
- Improved performance: Waypoint proxies can cache routing and policy information, which can improve performance for workloads in the namespace.

> - Increased security: Waypoint proxies can enforce authorization policies at the namespace level, which can help to protect your workloads from unauthorized access.

Reviewer: It is unclear what you are comparing with?  deploy ns vs sa scoped waypoint?

However, deploying Waypoint proxies at the namespace level also has some drawbacks, including:
- Limited granularity: You cannot apply different policies to different workloads in the same namespace.
- Increased resource consumption: Each namespace will require its own Waypoint proxy, which can consume more resources.

### Service account-level scope

Deploying Waypoint proxies at the service account level provides a number of benefits, including:
- Increased granularity: You can apply different policies to different workloads based on their service account.

> - Reduced resource consumption: You only need to deploy a Waypoint proxy for each service account that has workloads that require L7 routing or policy enforcement.

Reviewer: similar comment here - https://github.com/istio/istio.io/pull/13954/files#r1414070977

However, deploying Waypoint proxies at the service account level also has some drawbacks, including:

- Increased complexity: Managing policies at the service account level can be more complex, especially if you have a large number of service accounts.
- Reduced caching: Waypoint proxies cannot cache routing and policy information at the service account level, which can reduce performance.

### How to choose the right scope for your Waypoint proxies

The best scope for your Waypoint proxies will depend on your specific needs and requirements. If you have a simple application with a small number of workloads, then namespace-level scope may be a good choice. However, if you have a more complex application with a large number of workloads or if you need to apply different policies to different workloads, then service account-level scope may be a better choice.

Here are some factors to consider when choosing the scope for your Waypoint proxies:

> - The number of workloads in your application

Reviewer: Workloads sharing the same service account or not?

- The complexity of your application
- The need to apply different policies to different workloads
- The performance requirements of your application
- The resource requirements of your application

## Functional Overview

The functional behaviour of the waypoint proxy is dynamically configured by Istio to serve your applications configurations. This section takes a brief look at these functional aspects - detailed description of the internal design of the waypoint proxy is out of scope for this guide. The detailed functional overview from the Secure Overlay Networking was already discussed in the [Ztunnel L4 Networking Guide](https://deploy-preview-13635--preliminary-istio.netlify.app/latest/docs/ops/ambient/usage/ztunnel/#functionaloverview) hence this section only focuses on functionalities and features that Waypoint Proxy provides.

{{< image width="100%"
link="waypoint-architecture.png"
caption="Waypoint architecture"
>}}

> What is unique about the waypoint proxy is that it runs either per-namespace (default) or per-service account. By running outside of the application pod, a waypoint proxy can install, upgrade, and scale independently from the application pod, providing a centralized approach to managing L7 traffic and enforcing policies as well as reduce operational costs.

Reviewer: This should be mentioned earlier?

Upon deployment of a gateway resource using the `kubectl apply` command, Istio's control plane, IstioD, assumes the role of the Waypoint controller. Recognizing the gateway resource with the "istio.io/waypoint" gateway class name, istiod automatically deploys the Waypoint proxy based on the specified configuration in the gateway resource.

The waypoint proxy's data plane operates at Layer 7, enabling it to fully parse connections into individual requests and apply policies based on request properties such as headers and credentials. This granular control over L7 traffic extends to a comprehensive suite of capabilities, including:

> - HTTP 1.x, 2, or 3

Reviewer: HTTP 3 - we don't support it yet
